pkg/maps: add BPF ipcache for egress policy enforcement

ianvernon opened the pull request:
Plumb the contents of the ip-identity cache to a BPF map for lookup in the datapath.

Signed-off by: Ian Vernon <ian@cilium.io>

Related-to: #1488 

```release-note
Plumb the contents of the ip-identity cache to a BPF map for lookup in the datapath.
```

Commit:
pkg/maps: add BPF ipcache map

This new map maps IPs to their corresponding numerical security identities.
Whenever the IPCache is updated, a callback is made to update this new map
as well. There is only one such map per cilium-agent. This map is used
by the datapath to get the identity which corresponds to an IP of a packet
upon egress to make a decision whether to allow or deny a packet upon
egress.

Also starts the watcher for ip-identity in the key-value store after the map
is cleaned up in daemon.init(), otherwise the contents of the map will be
cleaned after the watcher starts, which will result in inconsistent
state in the agent.

Signed-off by: Ian Vernon <ian@cilium.io>

joestringer commented on pkg/identity/identity.go:
```diff
@@ -30,6 +32,17 @@ type Identity struct {
+//
+// WARNING - STABLE API
+// This structure is written as JSON to the key-value store. Do NOT modify this
+// structure in ways which are not JSON forward compatible.
```
I don't see JSON formatting expressions after the field declarations below.. should there be?

Commit:
pkg/bpf: log when we call DeleteAll() on map

Signed-off by: Ian Vernon <ian@cilium.io>

aanm commented on daemon/daemon.go:
```diff
@@ -835,6 +836,10 @@ func (d *Daemon) init() error {
 			return err
 		}
 
+		if err := remotelxcmap.RemoteLXCMap.DeleteAll(); err != nil {
```
Can you add a comment on what is this for?

aanm commented on pkg/maps/remotelxcmap/remotelxcmap.go:
```diff
@@ -0,0 +1,75 @@
+package remotelxcmap
```
Needs copywrite header

raybejjani commented on cilium/cmd/bpf_endpoint_ip_list.go:
```diff
@@ -0,0 +1,57 @@
+
+		bpfEndpointIpList := make(map[string][]string)
+		if err := remotelxcmap.RemoteLXCMap.Dump(bpfEndpointIpList); err != nil {
+			os.Exit(1)
```
Should these be fatal logs of the errors? I'm not sure if we normally `Exit(1)` with other commands but it can't hurt to print a real error.

ianvernon replied:
Error now printed in latest upload.

joestringer replied:
Strictly speaking, errors should probably be logged to stderr:

`fmt.Fprintf(os.Stderr, ...)`

raybejjani commented on daemon/endpoint.go:
```diff
@@ -681,7 +684,31 @@ func (h *putEndpointIDLabels) Handle(params PutEndpointIDLabelsParams) middlewar
+	key := remotelxcmap.NewEndpointKey(ipIDPair.IP)
+	value := remotelxcmap.RemoteEndpointInfo{SecurityIdentity: uint16(ipIDPair.ID)}
+
+	if modType == ipcache.Upsert {
```
a `switch` might be better suited here instead of the if-else chain

ianvernon replied:
Good idea, fixed in latest upload.

joestringer commented on cilium/cmd/bpf_endpoint_ip_list.go:
```diff
@@ -0,0 +1,57 @@
+
+const (
+	ipAddrTitle             = "IP ADDRESS"
+	remoteEndpointInfoTitle = "REMOTE ENDPOINT INFO"
```
Describing this in more concrete terms would be helpful. Specifically, the info is the `Identity`.

joestringer reviewed (approved):
Looks good to me. Minor tidyup suggestions.

tgraf commented on pkg/maps/remotelxcmap/remotelxcmap.go:
```diff
@@ -0,0 +1,89 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package remotelxcmap
```
I'm not too fond of the name remotelxcmap. This map should really not be related to lxc or endpoints at all. It maps ips to identities. Cann we just call it `ipcache`?

tgraf commented on daemon/endpoint.go:
```diff
@@ -681,7 +684,31 @@ func (h *putEndpointIDLabels) Handle(params PutEndpointIDLabelsParams) middlewar
+
+	switch modType {
+	case ipcache.Upsert:
+		err := remotelxcmap.RemoteLXCMap.Update(key, value)
```
Please add a FIXME referring https://github.com/cilium/cilium/issues/3161 so we fix this up. We need to get rid of these one time attempts with logging as the only fallback

tgraf commented on daemon/daemon.go:
```diff
@@ -835,6 +836,13 @@ func (d *Daemon) init() error {
+		// Clean all ip-identity entries from BPF map as watcher of key-value
+		// store from which these entries are obtained is started quickly
+		// after boot.
+		if err := remotelxcmap.RemoteLXCMap.DeleteAll(); err != nil {
```
I'm not sure this is safe, it means that a restart will temporarily  block all egress traffic. If you really want to do this, then mark the map `NonPersistent` so it will get removed and re-created on opening. It has the benefit that we can change the map format across Cilium versions.

ianvernon replied:
Upon startup, maybe we could have the agent set some flag on each value (some mark) that signifies that the entries in the map were there upon agent startup. When the IPIdentityWatcher tries to insert them into the map, it updates the value to not have the mark.  

I can change the IPIdentityWatcher to do an initial `ListPrefix`. Once the initial list from the key-value store is complete, we can then delete all keys that contain this marker by walking over the map again. 

This will handle cases where an endpoint might be deleted on a remote node while the local agent was down / restarted, and will ensure no disruption in the datapath.

ianvernon replied:
I've addressed all but this for now.

tgraf commented on pkg/maps/remotelxcmap/remotelxcmap.go:
```diff
@@ -0,0 +1,89 @@
+const (
+	// MaxKeys is the maximum number of keys that can be present in the
+	// RemoteEndpointMap.
+	MaxKeys = 0xFFFF
```
65535 is not enough as this is cluster wide. I suggest 512K

tgraf commented on cilium/cmd/bpf_endpoint_ip_list.go:
```diff
@@ -0,0 +1,57 @@
+	remoteEndpointInfoTitle = "REMOTE ENDPOINT INFO"
+)
+
+var bpfEndpointIpListCmd = &cobra.Command{
```
Let's not put this under endpoint, `bpf endpoint` is supposed to be for per endpoint data. Let's do `cilium bpf ipcache list`.

Commit:
bpf: rename cilium_remote_lxc to cilium_ipcache

Bump up amount of entries in map to 512000 as well.

Signed-off by: Ian Vernon <ian@cilium.io>
